lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs

```diff
@@ -0,0 +1,474 @@
+        /// </summary>
+        private readonly string _masterTaskId;
+
+        /// <summary>
+        /// Total number of Application error received from tasks
+        /// </summary>
+        private int _numberOfAppError = 0;
+
+        /// <summary>
+        /// Creates a TaskManager with specified total number of tasks and master task id.
+        /// Throws IMRUSystemException if numTasks is 0 or masterTaskId is null.
+        /// </summary>
+        /// <param name="numTasks"></param>
+        /// <param name="masterTaskId"></param>
+        internal TaskManager(int numTasks, string masterTaskId)
+        {
+            if (numTasks <= 0)
+            {
+                Exceptions.Throw(new IMRUSystemException("Number of expected task cannot be 0"), Logger);
+            }
+
+            if (string.IsNullOrWhiteSpace(masterTaskId))
+            {
+                Exceptions.Throw(new IMRUSystemException("masterTaskId cannot be null"), Logger);
+            }
+
+            _totalExpectedTasks = numTasks;
+            _masterTaskId = masterTaskId;
+        }
+
+        /// <summary>
+        /// Adds a Task to the task collection
+        /// Throws IMRUSystemException in the following cases:
+        ///   taskId is already added 
+        ///   taskConfiguration is null
+        ///   activeContext is null
+        ///   trying to add extra tasks
+        ///   trying to add Master Task twice
+        ///   No Master Task is added in the collection
+        /// </summary>
+        /// <param name="taskId"></param>
+        /// <param name="taskConfiguration"></param>
+        /// <param name="activeContext"></param>
+        internal void AddTask(string taskId, IConfiguration taskConfiguration, IActiveContext activeContext)
+        {
+            if (taskId == null)
+            {
+                Exceptions.Throw(new IMRUSystemException("The taskId is null."), Logger);
+            }
+
+            if (_tasks.ContainsKey(taskId))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The task [{0}] already exists.", taskId);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            if (taskConfiguration == null)
+            {
+                Exceptions.Throw(new IMRUSystemException("The task configuration is null."), Logger);
+            }
+
+            if (activeContext == null)
+            {
+                Exceptions.Throw(new IMRUSystemException("The context is null."), Logger);
+            }
+
+            if (NumberOfTasks >= _totalExpectedTasks)
+            {
+                string msg = string.Format("Trying to add an additional Task {0}, but the total expected Task number {1} has been reached.", taskId, _totalExpectedTasks);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            if (taskId.Equals(_masterTaskId) && MasterTaskExists())
```

Comment:
Is this code path reachable? Second master task is a task with the same id as already exists, so it should be covered by `if _tasks.ContainsKey(taskId)` branch. The test for it only checks that an exception of correct type is thrown, not the exception message